src/source-windivert.c

```diff
@@ -0,0 +1,682 @@
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    WinDivertQueueVars *wd_qv = WinDivertGetQueue(wd_tv->thread_num);
+
+    if (wd_qv == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "queue == NULL");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    SCMutexLock(&wd_qv->filter_init_mutex);
+    /* does the queue already have an active handle? */
+    if (wd_qv->filter_handle != NULL &&
+            wd_qv->filter_handle != INVALID_HANDLE_VALUE) {
+        goto unlock;
+    }
+
+    /* we open now so that we can immediately start handling packets, instead of
+     * losing however many would occur between registering the queue and
+     * starting a receive thread. */
+    wd_qv->filter_handle = WinDivertOpen(
+            wd_qv->filter_str, wd_qv->layer, wd_qv->priority, wd_qv->flags);
+    if (wd_qv->filter_handle == INVALID_HANDLE_VALUE) {
+        SCLogError(SC_ERR_FATAL, "WinDivertOpen failed, error: %" PRIu32 "",
+                (uint32_t)(GetLastError()));
+        ret = TM_ECODE_FAILED;
+        goto unlock;
+    }
+
+unlock:
+    if (ret == 0) { /* success */
+        wd_tv->filter_handle = wd_qv->filter_handle;
+
+        /* set our return context */
+        *data = wd_tv;
+    }
+
+    SCMutexUnlock(&wd_qv->filter_init_mutex);
+    SCReturnInt(ret);
+}
+
+/**
+ * \brief Deinit function releases resources at exit.
+ *
+ * \param tv pointer to generic thread vars
+ * \param data pointer to WinDivert-specific thread vars
+ */
+TmEcode ReceiveWinDivertThreadDeinit(ThreadVars *tv, void *data)
+{
+    SCEnter();
+
+    WinDivertThreadVars *wd_tv = (WinDivertThreadVars *)data;
+
+    SCReturnCT(WinDivertCloseHelper(wd_tv), "TmEcode");
+}
+
+/**
+ * \brief ExitStats prints stats to stdout at exit
+ *
+ *
+ * \param tv pointer to generic thread vars
+ * \param data pointer to WinDivert-specific thread vars
+ */
+void ReceiveWinDivertThreadExitStats(ThreadVars *tv, void *data)
+{
+    SCEnter();
+
+    WinDivertThreadVars *wd_tv = (WinDivertThreadVars *)data;
+    WinDivertQueueVars *wd_qv = WinDivertGetQueue(wd_tv->thread_num);
+    if (wd_qv == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "queue == NULL");
+        SCReturn;
+    }
+
+    SCMutexLock(&wd_qv->counters_mutex);
+
+    SCLogInfo("(%s) Packets %" PRIu32 ", Bytes %" PRIu64 ", Errors %" PRIu32 "",
+            tv->name, wd_qv->pkts, wd_qv->bytes, wd_qv->errs);
+    SCLogInfo("(%s) Verdict: Accepted %" PRIu32 ", Dropped %" PRIu32
+              ", Replaced %" PRIu32 "",
+            tv->name, wd_qv->accepted, wd_qv->dropped, wd_qv->replaced);
+
+    SCMutexUnlock(&wd_qv->counters_mutex);
+    SCReturn;
+}
+
+/**
+ * \brief WinDivert verdict module packet entry function
+ */
+TmEcode VerdictWinDivert(ThreadVars *tv, Packet *p, void *data, PacketQueue *pq,
+        PacketQueue *postpq)
+{
+    SCEnter();
+
+    TmEcode ret = TM_ECODE_OK;
+
+    /* \todo do we need to specifically handle tunnel packets like NFQ? */
```

Comment:
on receiving tunnel packets Suricata will split a packet into multiple. The encapsulated packets remain connected to the original packet (p->root). If we need to drop based on something detected in the encapsulated packet, we need to effectively drop the 'root' packet. This is what the tunnel handling does, so I'm pretty sure you'll need to handle it as well somehow.

Reply:
Thank you for the explanation. This situation should be handled properly now.